File: api-atom-java/main/java/net/aegistudio/api/java/ApiRemote.java
package net.aegistudio.api.java;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.function.BiFunction;

/**
 * Refers to a remote interface on server side, or a
 * remote callback on client side.
 * 
 * @author aegistudio
 */

public class ApiRemote {
	protected final int handle;
	protected final ApiHost host;
	
	protected ApiRemote(ApiHost host, int objectId) {
		this.handle = objectId;
		this.host = host;
	}
	
	public int handle() {
		return handle;
	}
	
	protected static <T extends ApiRemote> T read(
			DataInputStream dataInputStream, ApiHost apiHost, 
			BiFunction<ApiHost, Integer, T> factory) 
					throws IOException, ApiException {
		int result = dataInputStream.readInt();
		return result == 0? null : factory.apply(apiHost, result);
	}
	
	protected static void write(ApiRemote remote, 
			DataOutputStream dataOutputStream, ApiHost apiHost) 
					throws IOException, ApiException {
		dataOutputStream.writeInt(remote != null? remote.handle : 0);
	}
}

File: api-atom-cpp/src/apiRemote.cpp
#include "apiRemote.h"

using namespace api;

ApiRemote::ApiRemote():
	host(NULL), handle(0) { }

bool ApiRemote::nullPointer() const { 
	return handle == 0; 
}

ApiRemote::ApiRemote(const ApiRemote& _copy):
	host(_copy.host), handle(_copy.handle) {

}

_EX(variant<int8_t>) ApiRemote::call(
	int32_t call, variant<int8_t> parameter) {

	if(nullPointer()) throwException(
		"Invoke on null pointer!");
	return host -> call(handle, call, parameter);
}

void ApiRemote::read(ApiHost& _host, 
	InputStream& _inputStream) {

	host = &_host;
	handle = _inputStream.readInt();
}

void ApiRemote::write(ApiHost& _host,
	OutputStream& _outputStream) {

	if(host != &_host)
		_outputStream.writeInt(0);
	else _outputStream.writeInt(handle);
}
